keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,662 @@
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+Pod affinity/anti-affinity API allows using this feature across namespaces using 
+a static list, which works if the user knowns the namespaces names in advance.
+
+However, there are cases where the namespaces are not known beforehand, for which there
+is no way to use pod affinity/anti-affinity. Allowing users to specify the set of namespaces
+using a namespace selector addresses this problem.
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Allow users to dynamically select the set of namespaces considered when using pod
+  affinity/anti-affinity
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+<!--
+### Non-Goals
+
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+I am running a SaaS service, the workload for each customer is placed in a separate
+namespace. The workloads requires 1:1 pod to node placement. I want to use pod 
+anti-affinity across all customers namespaces to achieve that.
```

Comment:
Can you outline how this requirement would be achieved with the proposed change?

Presumably the namespaces are created on-the-fly for each customer so their names are not known ahead of time. But the labels on the namespaces would be known (say, something like stack=customer1, stack=customer2, etc.). So is the idea that for each customer's Pod somehow the PodAffinityTerm's Namespace selector would be specified specific to that customer in case of affinity? (say, for customer1's Pods it would be: 'stack=customer1') And for anti-affinity, it would be a list of all other labels (say, for customer1's Pods it would be: 'stack=customer2', 'stack=customer3', and so on).

Separately, you mention about 1:1 pod to node placement. Wouldn't this be possible with Pod's nodeSelector attribute? How is node placement related to namespace selection with regards to affinity/anti-affinity?

Reply:
> Can you outline how this requirement would be achieved with the proposed change?
> 
> Presumably the namespaces are created on-the-fly for each customer so their names are not known ahead of time. But the labels on the namespaces would be known (say, something like stack=customer1, stack=customer2, etc.). So is the idea that for each customer's Pods somehow the PodAffinityTerm's Namespace selector would be specified specific to that customer in case of affinity? (say, for customer1's Pods it would be: 'stack=customer1') And for anti-affinity, it would be a list of all other labels (say, for customer1's Pods it would be: 'stack=customer2', 'stack=customer3', and so on).

I am not quite following your example, but the idea is correct: while the namespaces may not be known in advance because they are created dynamically, they could be created with a common label. To achieve affinity/anti-affinity across those namespaces, the pods would be created with a term that uses a NamespaceSelector targeting the common namespace label.

> 
> Separately, you mention about 1:1 pod to node placement. Wouldn't this be possible with Pod's nodeSelector attribute? How is node placement related to namespace selection with regards to affinity/anti-affinity?

The component that creates the pods wouldn't know which nodes have capacity/assigned and which are not, OTOH the scheduler does and is able to pick the node that both has capacity and satisfies the affinity constraints (among other possible things). Moreover, you may want to achieve 1:1 not across all possible pods, but across a specific set (e.g., only one workload pod per node, but the node could host other control plane pods like daemonsets)

Reply:
Agree on the point that the Pod creating component may not know about node capacities.
Regarding achieving 1:1 mapping between workload pods and nodes, I guess anti-affinity rule with Namespace label selector will make it possible. Ok.

Reply:
> Regarding achieving 1:1 mapping between workload pods and nodes, I guess anti-affinity rule with Namespace label selector will make it possible. Ok.

Yes, exactly, and this is a fairly common case in HPC